### ISSUE TITLE
Remote Attestation Verification

### ISSUE THREAD

**Issue by pnguyen4** [oscarlab/graphene#1035]
I have a server that:
1. Recieves input from a remote client
2. Interprets input with application inside Graphene-SGX
3. Returns result to remote client

With the way remote attestation is currently implemented in Graphene there is no way for the remote client to see or verify the attestation process. As far as I can tell, the remote party in this scenario has no means to trust the response from the server. 

Would the devs be open to the idea of exporting either the quote or IAS report for independent verification? Are there any intentions to support a use case like this? If so, what would be the best way to approach this?

Thank you

EDIT
I found previous conversation that seems related to my original post. I'll leave it here.
https://github.com/oscarlab/graphene/pull/734#pullrequestreview-269558033

**Comment by pnguyen4** [comment]
I found previous conversation that seems related to my original post. I'll leave it here.
https://github.com/oscarlab/graphene/pull/734#pullrequestreview-269558033

**Comment by dimakuv** [comment]
Dear Phillip,

Thanks for the question. Indeed, Graphene-SGX currently has the *machinery* to perform SGX-based remote attestation (we merged this for the stable 1.0 release), but it is not yet available to the application/remote user. The goal was to have some foundation (AESM communication, IAS communication, helper functions) to build upon.

We have plans to incorporate the features similar to SGX-RA-TLS:
1. The attestation may be generated during run-time upon the application's request and with application-specific 64B of data (e.g., the hash of the public key exposed by the enclavized application.
2. The quote + the attestation report are exposed to the application. The application can now go and forward them to the remote user (in SGX-RA-TLS, this is baked into the normal TLS handshake protocol).

We will probably have other ways to expose remote attestation to the app/user.

TLDR: Yes, we have plans to expose the quote and the IAS attestation report to the application/remote user. This will land in the mainline in ~1-2 months (so, end of 2019).

**Comment by ArnaudCarrere** [comment]
Hi, do you have updates to share about this feature implementation ?
Thanks !

**Comment by dimakuv** [comment]
@ArnaudCarrere, there is an active development effort for this. It goes a bit slower that expected, but we'll hopefully have a draft PR in January and then merge it closer to February.

**Comment by dirknono** [comment]
Hi, 

Any updates on this RA feature?
I am currently trying to use Graphene at server side, and a Remote Attestation Verification feature would be very much helpful~

Thanks a lot,
Ray

**Comment by dimakuv** [comment]
@dirknono We have this PR: https://github.com/oscarlab/graphene/pull/1284. Still working on improving this PR and merging into master.

**Comment by mkow** [comment]
The requested feature is already on master (see #1383, #1392, #1414 and #1532), closing.
